### PR TITLE
Add a Sort ingest processor

### PR DESCRIPTION
**Pull request by polyfractal**
An ingest processor that sorts an array of values in ascending or descending order. 

If all elements are numerics, they will be sorted numerically. If values are strings, or mixtures of strings/numbers, the elements will be sorted lexicographically.

I started to implement the OpenRefine fingerprint stuff, but decided the individual components would be more useful as standalone processors (sort, dedupe, etc).  Thoughts?  If this isn't seen as useful individually, I'm happy to close and move this code into a single processor.
- ~~The sorting comparator gave me a really hard time.  I couldn't figure out a way to sort the generic list with a chained reverse without specifically setting it up per-type.  Would love to know if there is a better way, this is really gross!~~
- I used an enum for asc/desc, because type safety.  But enums are so verbose and clunky, maybe it wasn't worth it?

### CODE REVIEW

**Comment by rmuir** [comment]
You tried using https://docs.oracle.com/javase/7/docs/api/java/util/Collections.html#reverseOrder%28java.util.Comparator%29 ?

**Comment by polyfractal** [comment]
Doh.  That, and making the list's wildcard extend Comparable fixed my prior issues.  Thanks! :)

File: core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+
+        SortOrder order = randomBoolean() ? SortOrder.ASCENDING : SortOrder.DESCENDING;
+        if (order.equals(SortOrder.DESCENDING)) {
+            Collections.reverse(expectedResult);
+        }
+
+        String fieldName = RandomDocumentPicks.addRandomField(random(), ingestDocument, fieldValue);
+        Processor processor = new SortProcessor(randomAsciiOfLength(10), fieldName, order);
+        processor.execute(ingestDocument);
+        assertThat(ingestDocument.getFieldValue(fieldName, List.class), equalTo(expectedResult));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java]: I prefer `assertEquals` in cases like this. `assertThat` is great if you need to take a matcher or want to assert something complicated, but I like `assertEquals` for equality.

File: core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+
+    public void testSortMixedObjects() throws Exception {
+        IngestDocument ingestDocument = RandomDocumentPicks.randomIngestDocument(random());
+        int numItems = randomIntBetween(1, 10);
+        List<Object> fieldValue = new ArrayList<>(numItems);
+        List<Object> expectedResult = new ArrayList<>(numItems);
+        Object value;
+        for (int j = 0; j < numItems; j++) {
+            if (randomBoolean()) {
+                value = String.valueOf(randomIntBetween(0, 100));
```
Comment (nik9000) [inline core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java]: This seems pretty much the same as the test above modulo the type parameter of the list - but those aren't available at run time. How is it different?

Comment (polyfractal) [inline core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java]: Yeah, that was largely for my sanity.  I wasn't sure how the Ingest pipeline would deal with heterogeneous objects.  E.g. would it cast everything to a string, or throw an exception, or ?

Answer was cast to string.  We can probably remove the test. :)

Comment (nik9000) [inline core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java]: I think if you dropped the `String.valueof` from this line it'd test a different thing?

File: core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        String fieldName = RandomDocumentPicks.addRandomField(random(), ingestDocument, fieldValue);
+        Processor processor = new SortProcessor(randomAsciiOfLength(10), fieldName, order);
+        processor.execute(ingestDocument);
+        assertEquals(ingestDocument.getFieldValue(fieldName, List.class), expectedResult);
+    }
+
+
+
```
Comment (martijnvg) [inline core/src/test/java/org/elasticsearch/ingest/processor/SortProcessorTests.java]: maybe also add a test that verifies that we throw an error if a field isn't an array?

File: docs/reference/ingest/ingest-node.asciidoc [hunk 4]
```diff
@@ -1215,6 +1215,31 @@ Splits a field into an array using a separator character. Only works on string f
...
+numerically, while arrays of strings or heterogeneous arrays of strings + numbers will be sorted lexicographically.
+Throws an error when the field is not an array.
+
+[[sort-options]]
+.Sort Options
+[options="header"]
+|======
+| Name          | Required  | Default  | Description
+| `field`       | yes       | -        | The field to be separated
+| `order`   | no       | `"asc"`        | The sort order to use. Accepts `"asc"` or `"desc"`.
```
Comment (martijnvg) [inline docs/reference/ingest/ingest-node.asciidoc]: spacing?

File: core/src/main/java/org/elasticsearch/ingest/processor/SortProcessor.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Processor that sorts an array of items.
+ * Throws exception is the specified field is not an array.
+ */
+public final class SortProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "sort";
+    public static final ParseField FIELD = new ParseField("field");
```
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/ingest/processor/SortProcessor.java]: Maybe change the type to `String`? In all the other processors we use `String` too. We don't have any logic that deals with bwc. So I think until we add that to `ConfigurationUtils` we should just use String.

**Comment by martijnvg** [comment]
Left some minor comments. We should also add a rest integration test that verifies that this processor can actually be used in a pipeline on a real node. (this catches things like, forgetting to register the factory or if a processor requires special permissions that these permission are granted)
We do this also for all the other processors. (for example check the yaml test for the `foreach` processor).

Other than that this PR looks good!

File: core/src/main/java/org/elasticsearch/ingest/processor/SortProcessor.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    public String getType() {
+        return TYPE;
+    }
+
+    public final static class Factory extends AbstractProcessorFactory<SortProcessor> {
+
+        @Override
+        public SortProcessor doCreate(String processorTag, Map<String, Object> config) throws Exception {
+            String field = ConfigurationUtils.readStringProperty(TYPE, processorTag, config, FIELD.getPreferredName());
+            SortOrder direction = SortOrder.fromString(
```
Comment (talevy) [inline core/src/main/java/org/elasticsearch/ingest/processor/SortProcessor.java]: we should rethrow this potential `IllegalArgumentException` into a `ElasticsearchParseException` with the help of

 ```java
try {
  ...SortOrder.fromString...
} catch (IllegalArgumentException e) {
  ConfigurationUtils.newConfigurationException(TYPE, processorTag, ORDER.getPreferredName(), e.getMessage())
}
```

This way we will have a more structured error message during pipeline creation

**Comment by polyfractal** [comment]
@martijnvg Updated the PR based on comments while you were away, lemme know what you think when you get a chance (no rush though!) :)

File: docs/reference/ingest/ingest-node.asciidoc [hunk 7]
```diff
@@ -1282,6 +1282,31 @@ Splits a field into an array using a separator character. Only works on string f
...
+Sorts the elements of an array ascending or descending.  Homogeneous arrays of numbers will be sorted
+numerically, while arrays of strings or heterogeneous arrays of strings + numbers will be sorted lexicographically.
+Throws an error when the field is not an array.
+
+[[sort-options]]
+.Sort Options
+[options="header"]
+|======
+| Name    | Required  | Default  | Description
+| `field` | yes       | -        | The field to be separated
```
Comment (talevy) [inline docs/reference/ingest/ingest-node.asciidoc]: do we want this field to be templateable?

Comment (martijnvg) [inline docs/reference/ingest/ingest-node.asciidoc]: I think so. Also for other processors we should consider if the `field` option should be templateable. Maybe we should make the `field` options templateable via another change?

**Comment by martijnvg** [comment]
LGTM

**Pull request merged**